spec.md

```diff
@@ -0,0 +1,1311 @@
+    - [Basics of updating](#basics-of-updating-1)
+    - [Update statements](#update-statements-1)
+- [Pipelines](#pipelines)
+  - [Basics of streams](#basics-of-streams)
+  - [Clauses](#clauses)
+    - [Match](#match)
+    - [Insert](#insert)
+    - [Delete](#delete)
+    - [Update](#update)
+    - [Fetch](#fetch)
+  - [Operators](#operators)
+    - [Select](#select)
+    - [Sort](#sort)
+    - [Limit](#limit)
+    - [Offset](#offset)
+    - [Reduce](#reduce)
+  - [Execution](#execution)
+- [Transactionality and Concurrency](#transactionality-and-concurrency)
+  - [Basics](#basics)
+  - [Snapshots](#snapshots)
+  - [Isolation](#isolation)
+- [Sharding](#sharding)
+- [Glossary](#glossary)
+  - [Type system](#type-system)
+    - [Type](#type)
+    - [Schema type](#schema-type)
+    - [Value type](#value-type)
+    - [Data instance / instance](#data-instance--instance)
+    - [Data value / value](#data-value--value)
+    - [Attribute instance value / attribute value](#attribute-instance-value--attribute-value)
+    - [Data element / element](#data-element--element)
+    - [Concept](#concept)
+    - [Concept map](#concept-map)
+    - [Stream](#stream)
+    - [Answer set](#answer-set)
+    - [Answer](#answer)
+  - [TypeQL syntax](#typeql-syntax)
+    - [Schema query](#schema-query)
+    - [Data query](#data-query)
+    - [Clause / Stream clause](#clause--stream-clause)
+    - [Operators / Stream operator](#operators--stream-operator)
+    - [Functions](#functions)
+    - [Statement](#statement)
+    - [Pattern](#pattern)
+    - [Stream reduction / reduction](#stream-reduction--reduction)
+    - [Clause](#clause)
+    - [Block](#block)
+    - [Suffix](#suffix)
+
+
+# Foundations
+
+_(For reference only)_
+
+## Terminology
+
+* **TT** — transaction time
+  * _Interpretation_: any time within transaction
+* **CT** — commit time
+  * _Interpretation_: time of committing a transaction to DB
+* **tvar** — type variable
+* **evar** - data variable
+
+_Remark_: **tvar**s and **evar**s are uniquely distinguish everywhere in TypeQL
+
+## The type system
+
+* **Types**. We write 
+  $`A : \mathbf{Type}`$ to mean:
+  > $A$ is a type. 
+
+  * _Variations_: in general, may replace $\textbf{Type}$ by: 
+    * $\mathbf{Ent}$ (collection of entity types)
+    * $\mathbf{Rel}$ (collection of relation types)
+    * $`\mathbf{Att}`$ (collection of attribute types)
+    * $`\mathbf{Itf}`$ (collection of interface types)
+  * _Useful abbreviations_:
+    * $\mathbf{Obj} = \mathbf{Ent} + \mathbf{Rel}$ (collection of object types)
+    * $`\mathbf{ERA} = \mathbf{Ent} + \mathbf{Rel} + \mathbf{Att}`$ (collection of ERA types)
+  * _Example_: $`\mathsf{Person} : \mathbf{Ent}`$ means $`\mathsf{Person}`$ an entity type.
+* **Typing**
+  If $A$ is a type, then we may write $`a : A`$ to mean:
+  > $a$ is an element in type $A$.
+
+  * _Direct typing_: We write $a :_! A$ to mean:
+    > $a$ was declared as an element of $A$ by the user (we speak of a ***direct typing***).
```

Comment:
It is unclear what "direct" typing (or other "direct" things) is for and how it differs from regular typing. Perhaps this idea could be fleshed out a bit, or removed if unnecessary.